core/src/main/java/de/jplag/merging/MatchMerging.java

```diff
@@ -0,0 +1,136 @@
+        for (int i = 0; i < sortedByFirst.size() - 1; i++) {
+            if (sortedBySecond.indexOf(sortedByFirst.get(i)) == (sortedBySecond.indexOf(sortedByFirst.get(i + 1)) - 1)) {
+                neighbors.add(Arrays.asList(sortedByFirst.get(i), sortedByFirst.get(i + 1)));
+            }
+        }
+    }
+
+    private void mergeNeighbors() {
+        int i = 0;
+        while (i < neighbors.size()) {
+            int lengthUpper = neighbors.get(i).get(0).length();
+            int lengthLower = neighbors.get(i).get(1).length();
+            int seperatingFirst = neighbors.get(i).get(1).startOfFirst() - neighbors.get(i).get(0).endOfFirst() - 1;
+            int seperatingSecond = neighbors.get(i).get(1).startOfSecond() - neighbors.get(i).get(0).endOfSecond() - 1;
+            double seperating = (seperatingFirst + seperatingSecond) / 2.0;
+            // Checking length is not necessary as GST already checked length while computing matches
+            if (seperating <= seperatingThreshold) {
+                globalMatches.removeAll(neighbors.get(i));
+                globalMatches
+                        .add(new Match(neighbors.get(i).get(0).startOfFirst(), neighbors.get(i).get(0).startOfSecond(), lengthUpper + lengthLower));
+                removeToken(neighbors.get(i).get(0).startOfFirst(), neighbors.get(i).get(0).startOfSecond(), lengthUpper, seperatingFirst,
+                        seperatingSecond);
+                computeNeighbors();
+                i = 0;
+            } else {
+                i++;
+            }
+        }
+    }
+
+    private void removeToken(int startFirst, int startSecond, int lengthUpper, int seperatingFirst, int seperatingSecond) {
+        List<Token> tokenFirst = new ArrayList<>(firstSubmission.getTokenList());
+        List<Token> tokenSecond = new ArrayList<>(secondSubmission.getTokenList());
+        tokenFirst.subList(startFirst + lengthUpper, startFirst + lengthUpper + seperatingFirst).clear();
+        tokenSecond.subList(startSecond + lengthUpper, startSecond + lengthUpper + seperatingSecond).clear();
+        firstSubmission.setTokenList(tokenFirst);
+        secondSubmission.setTokenList(tokenSecond);
+
+        for (int i = 0; i < globalMatches.size(); i++) {
+            if (globalMatches.get(i).startOfFirst() > startFirst) {
+                Match alteredMatch = new Match(globalMatches.get(i).startOfFirst() - seperatingFirst, globalMatches.get(i).startOfSecond(),
+                        globalMatches.get(i).length());
+                globalMatches.set(i, alteredMatch);
+            }
+            if (globalMatches.get(i).startOfSecond() > startSecond) {
+                Match alteredMatch = new Match(globalMatches.get(i).startOfFirst(), globalMatches.get(i).startOfSecond() - seperatingSecond,
+                        globalMatches.get(i).length());
+                globalMatches.set(i, alteredMatch);
+            }
```

Comment:
For me personally, it hard to understand this section, as it is a bit difficult due to naming, lack of comments, and index magic. Why not add the changes to a new list of matches here?

Reply:
All it does, is move the starting positions of matches that are over the merge-point(is clear what I mean by this?) down by how many tokens were removed from the list.

I don't understand what your suggestion is.

Reply:
- Regarding naming: Adapt separatingFirst and separatingSecond
- Regarding comments: Your explanation here is could be a line comment, but you should explaining merge-point further.
- Regarding adding the changed matches to a list, see the global comment.

Reply:
Ah, thats what you meant